clj-parse.el

```diff
@@ -68,98 +88,147 @@
+     ((eq first-char ?o) (string-to-number (substring c 2) 8))
      (t first-char))))
 
-(defun clj-parse-edn-reduce1 (stack token)
-  (cl-case (cdr (assq 'type token))
-    (:whitespace stack)
-    (:number (cons (string-to-number (cdr (assq 'form token))) stack))
-    (:nil (cons nil stack))
-    (:true (cons t stack))
-    (:false (cons nil stack))
-    (:symbol (cons (intern (cdr (assq 'form token))) stack))
-    (:keyword (cons (intern (cdr (assq 'form token))) stack))
-    (:string (cons (clj-parse-string (cdr (assq 'form token))) stack))
-    (:character (cons (clj-parse-character (cdr (assq 'form token))) stack))))
-
-(defun clj-parse-edn-reduceN (stack type coll)
-  (if (eq :discard type)
-      stack
-    (cons
-     (cl-case type
-       (:whitespace :ws)
-       (:number coll)
-       (:list (-butlast (cdr coll)))
-       (:set (-butlast (cdr coll)))
-       (:vector (apply #'vector (-butlast (cdr coll))))
-       (:map (mapcar (lambda (pair)
-                       (cons (car pair) (cadr pair)))
-                     (-partition 2 (-butlast (cdr coll))))))
-     stack)))
-
-(defun clj-parse--reduce-coll (stack open-token coll-type reducN)
-  (let ((coll nil))
+(defun clj-parse--next ()
+  (setq next (clj-lex-next))
+  (while (eq (clj-lex-token-type next) :whitespace)
+    (setq next (clj-parse--next)))
+  next)
+
+(defun clj-parse--ast-reduce1 (stack leaf)
+  (push leaf stack))
+
+(defun clj-parse--ast-reduceN (stack node subnodes)
+  (push
+   (cl-case (clj-lex-token-type node)
+     (:lparen (clj-parse--make-node :list subnodes))
+     (:lbracket (clj-parse--make-node :vector subnodes))
+     (:set (clj-parse--make-node :set subnodes))
+     (:lbrace (clj-parse--make-node :map subnodes))
+     (:discard (clj-parse--make-node :discard subnodes)))
+   stack))
+
+(defun clj-parse--find-opener (stack closer-token)
+  (cl-case (clj-lex-token-type closer-token)
+    (:rparen :lparen)
+    (:rbracket :lbracket)
+    (:rbrace (clj-lex-token-type
+              (-find (lambda (token) (member (clj-lex-token-type token) '(:lbrace :set))) stack)))))
+
+(defun clj-parse--reduce-coll (stack closer-token reduceN)
+  "Reduce collection based on the top of the stack"
+  (let ((opener-type (clj-parse--find-opener stack closer-token))
+        (coll nil))
     (while (and stack
-                (not (eq (clj-lex-token-type (car stack)) open-token)))
+                (not (eq (clj-lex-token-type (car stack)) opener-type)))
       (push (pop stack) coll))
-    (if (eq (clj-lex-token-type (car stack)) open-token)
-        (progn
-          (push (pop stack) coll)
-          (funcall reduceN stack coll-type coll))
-      ;; Unwound the stack without finding a matching paren: return the original stack
-      (reverse list))))
+
+    (if (eq (clj-lex-token-type (car stack)) opener-type)
+        (let ((node (pop stack)))
+          (funcall reduceN stack node coll))
+      ;; Syntax error
+      (error "Syntax Error"))))
```

Comment:
yes, `reduce1` and `reduceN`, which maybe should be called `reduce-leaf` and `reduce-node`. Or even `shift-leaf` and `reduce-node`, since the first one will typically "shift" a single leaf node onto the stack, whereas the second one "reduces" a number of nodes into a single node.

you can think of parsing as three layers working together: the lexer, "the parser", "the reducer". The last part is pluggable to support different outputs.

`to-clojure-string` would be "the printer"